{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "NPfKJ3yBaPs5"
   },
   "source": [
    "This notebook prepares the data files so that they can be used for the training model. The training and test data frames are saved as pickles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "fDLhjIPdLKKx"
   },
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "colab": {
     "autoexec": {
      "startup": false,
      "wait_interval": 0
     }
    },
    "colab_type": "code",
    "id": "sRjzIy1ULO_w"
   },
   "outputs": [],
   "source": [
    "from sklearn.utils import shuffle\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import sys\n",
    "sys.path.append('/mnt/4_TB_HD/ramona/utils')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "oQ710RwpUBR8"
   },
   "source": [
    "# Modeling the text"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "b9i2Aw6wCgs9"
   },
   "source": [
    "## Transform the Training Set into Right Format \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {
     "autoexec": {
      "startup": false,
      "wait_interval": 0
     }
    },
    "colab_type": "code",
    "id": "7QXwn7aBKnpI"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training Variants (3321, 4)\n",
      "Train Text      (3321, 2)\n",
      "   ID    Gene             Variation  Class\n",
      "0   0  FAM58A  Truncating Mutations      1\n",
      "1   1     CBL                 W802*      2\n",
      "2   2     CBL                 Q249E      2\n",
      "3   3     CBL                 N454D      3\n",
      "4   4     CBL                 L399V      4\n",
      "(3321, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Recent evidence has demonstrated that acquired...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Oncogenic mutations in the monomeric Casitas B...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID                                               Text\n",
       "0   0  Cyclin-dependent kinases (CDKs) regulate a var...\n",
       "1   1   Abstract Background  Non-small cell lung canc...\n",
       "2   2   Abstract Background  Non-small cell lung canc...\n",
       "3   3  Recent evidence has demonstrated that acquired...\n",
       "4   4  Oncogenic mutations in the monomeric Casitas B..."
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "training_variants_df = pd.read_csv(\"../utils/data/training_variants\")\n",
    "training_text_df = pd.read_csv(\"../utils/data/training_text\", sep=\"\\|\\|\", engine=\"python\", skiprows=1, names=[\"ID\", \"Text\"])\n",
    "print(\"Training Variants\".ljust(15), training_variants_df.shape)\n",
    "print(\"Train Text\".ljust(15), training_text_df.shape)\n",
    "\n",
    "print(training_variants_df.head())\n",
    "print(training_text_df.shape)\n",
    "training_text_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate_classes(classe):\n",
    "    if int(classe) == 1:\n",
    "        meaning = 'Likely Loss-of-function'\n",
    "    elif int(classe) == 2:\n",
    "        meaning = 'Likely Gain-of-function'\n",
    "    elif int(classe) == 3:\n",
    "        meaning = 'Neutral'\n",
    "    elif int(classe) == 5:\n",
    "        meaning = 'Likely Neutral'\n",
    "    elif int(classe) == 6:\n",
    "        meaning = 'Inconclusive'\n",
    "    elif int(classe) == 7:\n",
    "        meaning = 'Gain-of-function'\n",
    "    elif int(classe) == 8:\n",
    "        meaning = 'Likely Switch-of-function'\n",
    "    elif int(classe) == 9:\n",
    "        meaning = 'Switch-of-function'\n",
    "    elif int(classe) == 4:\n",
    "        meaning = 'Loss-of-function'\n",
    "    return meaning\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "merge training text and variants into one data frame and drop the texts which have no content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "colab": {
     "autoexec": {
      "startup": false,
      "wait_interval": 0
     }
    },
    "colab_type": "code",
    "id": "P_l4dWtKXB0y"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3321, 5)\n",
      "        ID    Gene             Variation  Class Text  Text_count\n",
      "1109  1109   FANCA                S1088F      1  NaN           1\n",
      "1277  1277  ARID5B  Truncating Mutations      1  NaN           1\n",
      "1407  1407   FGFR3                 K508M      6  NaN           1\n",
      "1639  1639    FLT1         Amplification      6  NaN           1\n",
      "2755  2755    BRAF                 G596C      7  NaN           1\n",
      "(3316, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "      <th>Text</th>\n",
       "      <th>Text_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>FAM58A</td>\n",
       "      <td>Truncating Mutations</td>\n",
       "      <td>1</td>\n",
       "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
       "      <td>6089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>CBL</td>\n",
       "      <td>W802*</td>\n",
       "      <td>2</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "      <td>5756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>CBL</td>\n",
       "      <td>Q249E</td>\n",
       "      <td>2</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "      <td>5756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>CBL</td>\n",
       "      <td>N454D</td>\n",
       "      <td>3</td>\n",
       "      <td>Recent evidence has demonstrated that acquired...</td>\n",
       "      <td>5572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>CBL</td>\n",
       "      <td>L399V</td>\n",
       "      <td>4</td>\n",
       "      <td>Oncogenic mutations in the monomeric Casitas B...</td>\n",
       "      <td>6202</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID    Gene             Variation  Class  \\\n",
       "0   0  FAM58A  Truncating Mutations      1   \n",
       "1   1     CBL                 W802*      2   \n",
       "2   2     CBL                 Q249E      2   \n",
       "3   3     CBL                 N454D      3   \n",
       "4   4     CBL                 L399V      4   \n",
       "\n",
       "                                                Text  Text_count  \n",
       "0  Cyclin-dependent kinases (CDKs) regulate a var...        6089  \n",
       "1   Abstract Background  Non-small cell lung canc...        5756  \n",
       "2   Abstract Background  Non-small cell lung canc...        5756  \n",
       "3  Recent evidence has demonstrated that acquired...        5572  \n",
       "4  Oncogenic mutations in the monomeric Casitas B...        6202  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "training_merge_df = training_variants_df.merge(training_text_df,left_on=\"ID\",right_on=\"ID\")\n",
    "print(training_merge_df.shape)\n",
    "\n",
    "training_merge_df.loc[:, 'Text_count'] = training_merge_df[\"Text\"].apply(lambda x: len(str(x).split()) )\n",
    "print(training_merge_df.loc[training_merge_df['Text_count'] == 1])\n",
    "training_merge_df = training_merge_df[training_merge_df['Text_count'] != 1]\n",
    "\n",
    "#training_merge_df['Class'] = training_merge_df[\"Class\"].apply(lambda x: translate_classes(x) )\n",
    "print(training_merge_df.shape)\n",
    "#training_merge_df = shuffle(training_merge_df)\n",
    "training_merge_df.head() \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "#training_merge_df.to_csv('training_data_presentation.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "training_merge_df[['Gene', 'Variation']].to_pickle('test_variants_filtered')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There were 5 ID which had no text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "training_merge_df['Text'] = training_merge_df['Text'].apply(lambda x: x.replace('\\t',' '))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Saving the text of the training_merge_df combined with the class as a txt file for evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#training_merge_df[['Class','Text']].to_csv(r'training_text_with_label.txt', header=None, index=None, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "colab": {
     "autoexec": {
      "startup": false,
      "wait_interval": 0
     }
    },
    "colab_type": "code",
    "id": "mSDvvfEhTbbF"
   },
   "outputs": [],
   "source": [
    "def extract_text_sections(Text, Gene, Variation, ID): \n",
    "    section=''\n",
    "    index = Text.find(Variation)\n",
    "    if index == -1:\n",
    "      #print(\"Could not find: %s in iD: %d\"  % (Variation, ID))\n",
    "      Variation = 'mutation'\n",
    "      index = Text.find(Variation)\n",
    "    end_index= 0    \n",
    "    t= 0\n",
    "    #print(Variation,index,t)\n",
    "    index_list = []\n",
    "    while index != -1:\n",
    "      index_list.append(index)\n",
    "      old_index = index\n",
    "      index = Text.find(Variation, old_index+1)\n",
    "      t+=1\n",
    "    \n",
    "    \n",
    "    # for the case the variation appears less the 5 times\n",
    "    if t < 5 and t>0: \n",
    "  \n",
    "      for index in index_list: \n",
    "        # if two sections are overlapping \n",
    "        if index <= end_index: \n",
    "          # determine end point depending on if it's out of range or not \n",
    "          if index+int(2000/t) <= (len(Text)-1): \n",
    "            end = index+int(2000/t)\n",
    "          else: \n",
    "            end = (len(Text)-1)\n",
    "          section += Text[end_index:end]\n",
    "       \n",
    "        else: \n",
    "          # determine end point depending on if it's out of range or not\n",
    "          if index+int(2000/t) <=(len(Text)-1): \n",
    "            end = index+int(2000/t)\n",
    "          else: \n",
    "            end = (len(Text)-1)\n",
    "          # determine start point depending on if it's out of range or not\n",
    "          if index-int(2000/t) >= 0: \n",
    "            start = index-int(2000/t)\n",
    "          else: \n",
    "            start = 0\n",
    "          section += Text[start:end] \n",
    "         \n",
    "  \n",
    "        end_index = end\n",
    "    if t==0:\n",
    "      section = Text\n",
    "    else: \n",
    "       for index in index_list :\n",
    "        if index <= end_index: \n",
    "          # determine end point \n",
    "          if index+1000 <= (len(Text)-1): \n",
    "            end = index+1000\n",
    "          else: \n",
    "            end = (len(Text)-1)\n",
    "          section += Text[end_index:end]\n",
    "       \n",
    "        else: \n",
    "          if index+1000 <= (len(Text)-1): \n",
    "            end = index+1000\n",
    "          else: \n",
    "            end = (len(Text)-1)\n",
    "          if index-1000 >= 0: \n",
    "            start = index-1000\n",
    "          else: \n",
    "            start = 0\n",
    "          section += Text[start:end] \n",
    "        \n",
    "        end_index = end\n",
    "    section = section.replace('\\t', ' ')\n",
    "    return section\n",
    "   \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "      <th>Text</th>\n",
       "      <th>Text_count</th>\n",
       "      <th>modified_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>FAM58A</td>\n",
       "      <td>Truncating Mutations</td>\n",
       "      <td>1</td>\n",
       "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
       "      <td>6089</td>\n",
       "      <td>cyclin-dependent kinases (cdks) regulate a var...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>CBL</td>\n",
       "      <td>W802*</td>\n",
       "      <td>2</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "      <td>5756</td>\n",
       "      <td>ncer (nsclc) is a heterogeneous group of disor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>CBL</td>\n",
       "      <td>Q249E</td>\n",
       "      <td>2</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "      <td>5756</td>\n",
       "      <td>ll lung cancer (nsclc) is a heterogeneous grou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>CBL</td>\n",
       "      <td>N454D</td>\n",
       "      <td>3</td>\n",
       "      <td>Recent evidence has demonstrated that acquired...</td>\n",
       "      <td>5572</td>\n",
       "      <td>alysis but failed to detect any further sequen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>CBL</td>\n",
       "      <td>L399V</td>\n",
       "      <td>4</td>\n",
       "      <td>Oncogenic mutations in the monomeric Casitas B...</td>\n",
       "      <td>6202</td>\n",
       "      <td>d) compared to either a549 or hek293t cells (m...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID    Gene             Variation  Class  \\\n",
       "0   0  FAM58A  Truncating Mutations      1   \n",
       "1   1     CBL                 W802*      2   \n",
       "2   2     CBL                 Q249E      2   \n",
       "3   3     CBL                 N454D      3   \n",
       "4   4     CBL                 L399V      4   \n",
       "\n",
       "                                                Text  Text_count  \\\n",
       "0  Cyclin-dependent kinases (CDKs) regulate a var...        6089   \n",
       "1   Abstract Background  Non-small cell lung canc...        5756   \n",
       "2   Abstract Background  Non-small cell lung canc...        5756   \n",
       "3  Recent evidence has demonstrated that acquired...        5572   \n",
       "4  Oncogenic mutations in the monomeric Casitas B...        6202   \n",
       "\n",
       "                                       modified_text  \n",
       "0  cyclin-dependent kinases (cdks) regulate a var...  \n",
       "1  ncer (nsclc) is a heterogeneous group of disor...  \n",
       "2  ll lung cancer (nsclc) is a heterogeneous grou...  \n",
       "3  alysis but failed to detect any further sequen...  \n",
       "4  d) compared to either a549 or hek293t cells (m...  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "training_merge_df['modified_text'] = training_merge_df[['Text', 'Gene','Variation', 'ID']].apply(lambda x: extract_text_sections(x[0].lower(),x[1].lower(),x[2].lower(),x[3]), axis=1 )\n",
    "\n",
    "training_merge_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "colab": {
     "autoexec": {
      "startup": false,
      "wait_interval": 0
     }
    },
    "colab_type": "code",
    "id": "e4mb55sYi8b0"
   },
   "outputs": [],
   "source": [
    "training_merge_df.to_pickle('train_1000.sav')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "colab": {
     "autoexec": {
      "startup": false,
      "wait_interval": 0
     }
    },
    "colab_type": "code",
    "id": "nu6-JRYKRxKZ"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    1\n",
       "1    2\n",
       "2    2\n",
       "3    3\n",
       "4    4\n",
       "Name: Class, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Y_train = training_merge_df['Class']\n",
    "Y_train.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train.to_pickle('X_train_full.sav')\n",
    "Y_train.to_pickle('Y_train_full.sav')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "YRanlIpiEVBX"
   },
   "source": [
    "## Transform the Test Set into the Right Format "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {
     "autoexec": {
      "startup": false,
      "wait_interval": 0
     }
    },
    "colab_type": "code",
    "id": "YAChCEHJEKxO"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test Variant    (5668, 3)\n",
      "Test Text       (5668, 2)\n",
      "Solution        (368, 10)\n",
      "   ID  class1  class2  class3  class4  class5  class6  class7  class8  class9\n",
      "0  12       1       0       0       0       0       0       0       0       0\n",
      "1  19       0       1       0       0       0       0       0       0       0\n",
      "2  21       0       1       0       0       0       0       0       0       0\n",
      "3  55       0       0       0       1       0       0       0       0       0\n",
      "4  64       0       0       0       1       0       0       0       0       0\n",
      "Test Merge      (5668, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ACSL4</td>\n",
       "      <td>R570S</td>\n",
       "      <td>2. This mutation resulted in a myeloproliferat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>NAGLU</td>\n",
       "      <td>P521L</td>\n",
       "      <td>Abstract The Large Tumor Suppressor 1 (LATS1)...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>PAH</td>\n",
       "      <td>L333F</td>\n",
       "      <td>Vascular endothelial growth factor receptor (V...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ING1</td>\n",
       "      <td>A148D</td>\n",
       "      <td>Inflammatory myofibroblastic tumor (IMT) is a ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>TMEM216</td>\n",
       "      <td>G77A</td>\n",
       "      <td>Abstract Retinoblastoma is a pediatric retina...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID     Gene Variation                                               Text\n",
       "0   0    ACSL4     R570S  2. This mutation resulted in a myeloproliferat...\n",
       "1   1    NAGLU     P521L   Abstract The Large Tumor Suppressor 1 (LATS1)...\n",
       "2   2      PAH     L333F  Vascular endothelial growth factor receptor (V...\n",
       "3   3     ING1     A148D  Inflammatory myofibroblastic tumor (IMT) is a ...\n",
       "4   4  TMEM216      G77A   Abstract Retinoblastoma is a pediatric retina..."
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load test files\n",
    "\n",
    "test_variants_df = pd.read_csv(\"../utils/data/test_variants.csv\")\n",
    "test_text_df = pd.read_csv(\"../utils/data/test_text.csv\", sep=\"\\|\\|\", engine=\"python\", skiprows=1, names=[\"ID\", \"Text\"])\n",
    "solution = pd.read_csv(\"../utils/data/solution_filtered.csv\")\n",
    "print(\"Test Variant\".ljust(15), test_variants_df.shape)\n",
    "print(\"Test Text\".ljust(15), test_text_df.shape)\n",
    "print(\"Solution\".ljust(15), solution.shape)\n",
    "print(solution.head(5))\n",
    "\n",
    "test_merge_df = test_variants_df.merge(test_text_df,left_on=\"ID\",right_on=\"ID\")\n",
    "print('Test Merge'.ljust(15),test_merge_df.shape)\n",
    "test_merge_df.head(5)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "colab": {
     "autoexec": {
      "startup": false,
      "wait_interval": 0
     }
    },
    "colab_type": "code",
    "id": "Pi1k5vVr2Qz4"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>55</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>64</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID  Class\n",
       "0  12      1\n",
       "1  19      2\n",
       "2  21      2\n",
       "3  55      4\n",
       "4  64      4"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#build Y_test with the solution data frame \n",
    "\n",
    "solution.columns = ['ID', 1, 2, 3, 4, 5, 6,\n",
    "       7, 8, 9]\n",
    "Class_df = pd.DataFrame()\n",
    "Class_df['ID'] = solution['ID']\n",
    "Class_df['Class'] =  solution[[1, 2, 3, 4, 5, 6,\n",
    "       7, 8, 9]].idxmax(axis=1)\n",
    "\n",
    "Class_df.head(5) #Takes the info from solution\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "colab": {
     "autoexec": {
      "startup": false,
      "wait_interval": 0
     }
    },
    "colab_type": "code",
    "id": "u2YAbl2_4Kvh"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test classified (368, 5)\n",
      "       ID   Gene      Variation Text  Class  Text_count\n",
      "105  1623  AURKB  Amplification  NaN      2           1\n",
      "Test classified (367, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Text</th>\n",
       "      <th>Class</th>\n",
       "      <th>Text_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12</td>\n",
       "      <td>TET2</td>\n",
       "      <td>Y1902A</td>\n",
       "      <td>TET proteins oxidize 5-methylcytosine (5mC) on...</td>\n",
       "      <td>1</td>\n",
       "      <td>7123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>D2512H</td>\n",
       "      <td>Genes encoding components of the PI3K-Akt-mTOR...</td>\n",
       "      <td>2</td>\n",
       "      <td>4283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21</td>\n",
       "      <td>KIT</td>\n",
       "      <td>D52N</td>\n",
       "      <td>Myeloproliferative disorders (MPD) constitute ...</td>\n",
       "      <td>2</td>\n",
       "      <td>2477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>55</td>\n",
       "      <td>SPOP</td>\n",
       "      <td>F125V</td>\n",
       "      <td>In the largest E3 ligase subfamily, Cul3 binds...</td>\n",
       "      <td>4</td>\n",
       "      <td>22727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>64</td>\n",
       "      <td>KEAP1</td>\n",
       "      <td>C23Y</td>\n",
       "      <td>Keap1 is the substrate recognition module of a...</td>\n",
       "      <td>4</td>\n",
       "      <td>2854</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID   Gene Variation                                               Text  \\\n",
       "0  12   TET2    Y1902A  TET proteins oxidize 5-methylcytosine (5mC) on...   \n",
       "1  19   MTOR    D2512H  Genes encoding components of the PI3K-Akt-mTOR...   \n",
       "2  21    KIT      D52N  Myeloproliferative disorders (MPD) constitute ...   \n",
       "3  55   SPOP     F125V  In the largest E3 ligase subfamily, Cul3 binds...   \n",
       "4  64  KEAP1      C23Y  Keap1 is the substrate recognition module of a...   \n",
       "\n",
       "   Class  Text_count  \n",
       "0      1        7123  \n",
       "1      2        4283  \n",
       "2      2        2477  \n",
       "3      4       22727  \n",
       "4      4        2854  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_classified_df = test_merge_df.merge(Class_df,left_on=\"ID\",right_on=\"ID\")\n",
    "print('Test classified'.ljust(15),test_classified_df.shape)\n",
    "\n",
    "test_classified_df.loc[:, 'Text_count'] = test_classified_df[\"Text\"].apply(lambda x: len(str(x).split()) )\n",
    "print(test_classified_df.loc[test_classified_df['Text_count'] == 1])\n",
    "test_classified_df = test_classified_df[test_classified_df['Text_count'] != 1]\n",
    "\n",
    "print('Test classified'.ljust(15),test_classified_df.shape)\n",
    "test_classified_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_classified_df['modified_text'] = test_classified_df[['Text', 'Gene','Variation', 'ID']].apply(lambda x: extract_text_sections(x[0].lower(),x[1].lower(),x[2].lower(),x[3]), axis=1 )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Text</th>\n",
       "      <th>Class</th>\n",
       "      <th>Text_count</th>\n",
       "      <th>modified_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12</td>\n",
       "      <td>TET2</td>\n",
       "      <td>Y1902A</td>\n",
       "      <td>TET proteins oxidize 5-methylcytosine (5mC) on...</td>\n",
       "      <td>1</td>\n",
       "      <td>7123</td>\n",
       "      <td>the binding of purified tet2 proteins to a fa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>D2512H</td>\n",
       "      <td>Genes encoding components of the PI3K-Akt-mTOR...</td>\n",
       "      <td>2</td>\n",
       "      <td>4283</td>\n",
       "      <td>red somatic point mutations (supplementary tab...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21</td>\n",
       "      <td>KIT</td>\n",
       "      <td>D52N</td>\n",
       "      <td>Myeloproliferative disorders (MPD) constitute ...</td>\n",
       "      <td>2</td>\n",
       "      <td>2477</td>\n",
       "      <td>myeloproliferative disorders (mpd) constitute ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>55</td>\n",
       "      <td>SPOP</td>\n",
       "      <td>F125V</td>\n",
       "      <td>In the largest E3 ligase subfamily, Cul3 binds...</td>\n",
       "      <td>4</td>\n",
       "      <td>22727</td>\n",
       "      <td>re molecular signatures database (msigdb, cont...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>64</td>\n",
       "      <td>KEAP1</td>\n",
       "      <td>C23Y</td>\n",
       "      <td>Keap1 is the substrate recognition module of a...</td>\n",
       "      <td>4</td>\n",
       "      <td>2854</td>\n",
       "      <td>keap1 is the substrate recognition module of a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>70</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>G35V</td>\n",
       "      <td>Inherited mutations affecting the INK4a/ARF lo...</td>\n",
       "      <td>4</td>\n",
       "      <td>12660</td>\n",
       "      <td>inherited mutations affecting the ink4a/arf lo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>89</td>\n",
       "      <td>IDH1</td>\n",
       "      <td>R132S</td>\n",
       "      <td>To assess the value of anti-isocitrate dehydro...</td>\n",
       "      <td>9</td>\n",
       "      <td>27549</td>\n",
       "      <td>09 was generally crisper with a better signal-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>93</td>\n",
       "      <td>KIT</td>\n",
       "      <td>N822K</td>\n",
       "      <td>c-KIT is a member of the type 3 subclass of tr...</td>\n",
       "      <td>7</td>\n",
       "      <td>34300</td>\n",
       "      <td>us 10% fetal calf serum (fcs), and 4 mg cell l...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>111</td>\n",
       "      <td>ETV1</td>\n",
       "      <td>Amplification</td>\n",
       "      <td>E26 transformation-specific (ETS) transcriptio...</td>\n",
       "      <td>7</td>\n",
       "      <td>7082</td>\n",
       "      <td>to be fully explored. we have completed a com...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>113</td>\n",
       "      <td>JAK1</td>\n",
       "      <td>F958C</td>\n",
       "      <td>Activating mutations in JAK1 and JAK2 have bee...</td>\n",
       "      <td>7</td>\n",
       "      <td>4878</td>\n",
       "      <td>pocket formed between the jh2 αc helix and an ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>119</td>\n",
       "      <td>KIT</td>\n",
       "      <td>P551_M552del</td>\n",
       "      <td>The pathogenesis of most gastrointestinal stro...</td>\n",
       "      <td>2</td>\n",
       "      <td>3997</td>\n",
       "      <td>the pathogenesis of most gastrointestinal stro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>121</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>T1773I</td>\n",
       "      <td>Genetic screening of the breast and ovarian ca...</td>\n",
       "      <td>1</td>\n",
       "      <td>5245</td>\n",
       "      <td>lysis to assess the protein folding stability ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>126</td>\n",
       "      <td>TGFBR2</td>\n",
       "      <td>E526Q</td>\n",
       "      <td>Transforming growth factor β (TGFβ) plays a ke...</td>\n",
       "      <td>4</td>\n",
       "      <td>9016</td>\n",
       "      <td>β) plays a key role in maintaining tissue home...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>137</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Q144P</td>\n",
       "      <td>Inheritance of germ-line mutant alleles of BRC...</td>\n",
       "      <td>1</td>\n",
       "      <td>4635</td>\n",
       "      <td>inheritance of germ-line mutant alleles of brc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>159</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>Truncating Mutations</td>\n",
       "      <td>Abstract Heterozygous germ line mutations in ...</td>\n",
       "      <td>1</td>\n",
       "      <td>25258</td>\n",
       "      <td>ers of the family (two sisters) who were alrea...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>215</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>G173C</td>\n",
       "      <td>Exonic variants of unknown biological signific...</td>\n",
       "      <td>1</td>\n",
       "      <td>5883</td>\n",
       "      <td>exonic variants of unknown biological signific...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>226</td>\n",
       "      <td>GNAQ</td>\n",
       "      <td>R183Q</td>\n",
       "      <td>This study represents the most comprehensive m...</td>\n",
       "      <td>2</td>\n",
       "      <td>3439</td>\n",
       "      <td>ell et al., 1989; luyten et al., 1996; manioti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>234</td>\n",
       "      <td>MTOR</td>\n",
       "      <td>R2152C</td>\n",
       "      <td>Genes encoding components of the PI3K-Akt-mTOR...</td>\n",
       "      <td>3</td>\n",
       "      <td>4283</td>\n",
       "      <td>parental or wt-expressing cells (figure 3b an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>286</td>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>E453Q</td>\n",
       "      <td>Mutations in the PIK3CA gene, which encodes th...</td>\n",
       "      <td>2</td>\n",
       "      <td>6877</td>\n",
       "      <td>0a expression vector pcmv6-ish2p110a were kind...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>344</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>W559_R560del</td>\n",
       "      <td>Gastrointestinal stromal tumors (GISTs) are co...</td>\n",
       "      <td>7</td>\n",
       "      <td>1836</td>\n",
       "      <td>gastrointestinal stromal tumors (gists) are co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>353</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>S720P</td>\n",
       "      <td>The epidermal growth factor receptor tyrosine ...</td>\n",
       "      <td>2</td>\n",
       "      <td>6880</td>\n",
       "      <td>the epidermal growth factor receptor tyrosine ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>381</td>\n",
       "      <td>PPP2R1A</td>\n",
       "      <td>R183W</td>\n",
       "      <td>Protein Phosphatase 2A (PP2A) plays an essenti...</td>\n",
       "      <td>1</td>\n",
       "      <td>12291</td>\n",
       "      <td>s, the pp2a core enzyme associates with a vari...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>382</td>\n",
       "      <td>EWSR1</td>\n",
       "      <td>EWSR1-ERG Fusion</td>\n",
       "      <td>Ets-related gene (ERG), which encodes a member...</td>\n",
       "      <td>7</td>\n",
       "      <td>9050</td>\n",
       "      <td>ets-related gene (erg), which encodes a member...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>426</td>\n",
       "      <td>ERBB4</td>\n",
       "      <td>D931Y</td>\n",
       "      <td>Recent efforts to comprehensively characterize...</td>\n",
       "      <td>7</td>\n",
       "      <td>5337</td>\n",
       "      <td>recent efforts to comprehensively characterize...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>469</td>\n",
       "      <td>TSC1</td>\n",
       "      <td>R509Q</td>\n",
       "      <td>Tuberous sclerosis complex (TSC) is an autosom...</td>\n",
       "      <td>3</td>\n",
       "      <td>3820</td>\n",
       "      <td>y the complex and on the phosphorylation statu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>535</td>\n",
       "      <td>FLT3</td>\n",
       "      <td>D835V</td>\n",
       "      <td>Mutations of receptor tyrosine kinases are imp...</td>\n",
       "      <td>7</td>\n",
       "      <td>10694</td>\n",
       "      <td>mutations of receptor tyrosine kinases are imp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>542</td>\n",
       "      <td>SF3B1</td>\n",
       "      <td>H662D</td>\n",
       "      <td>The splicing factor SF3B1 is the most commonly...</td>\n",
       "      <td>9</td>\n",
       "      <td>5517</td>\n",
       "      <td>the splicing factor sf3b1 is the most commonly...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>545</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>F595L</td>\n",
       "      <td>Recently, mutations in the B-Raf gene have bee...</td>\n",
       "      <td>7</td>\n",
       "      <td>11983</td>\n",
       "      <td>recently, mutations in the b-raf gene have bee...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>552</td>\n",
       "      <td>PIK3R1</td>\n",
       "      <td>Truncating Mutations</td>\n",
       "      <td>Class IA phosphatidylinositol 3-kinases (PI3K)...</td>\n",
       "      <td>1</td>\n",
       "      <td>15901</td>\n",
       "      <td>class ia phosphatidylinositol 3-kinases (pi3k)...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>562</td>\n",
       "      <td>TP53</td>\n",
       "      <td>E180K</td>\n",
       "      <td>Under conditions of genotoxic stress, human p5...</td>\n",
       "      <td>4</td>\n",
       "      <td>12107</td>\n",
       "      <td>rs35,36. we also mutated corresponding interac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>5153</td>\n",
       "      <td>SMO</td>\n",
       "      <td>W281C</td>\n",
       "      <td>Introduction Uncontrolled activation of the He...</td>\n",
       "      <td>7</td>\n",
       "      <td>13821</td>\n",
       "      <td>the pocket without directly contacting the in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>5178</td>\n",
       "      <td>TP53</td>\n",
       "      <td>C176R</td>\n",
       "      <td>More than half of all human cancers are associ...</td>\n",
       "      <td>4</td>\n",
       "      <td>3069</td>\n",
       "      <td>mi) for the macintosh. the core domains ofadh-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>5192</td>\n",
       "      <td>FGFR2</td>\n",
       "      <td>N549H</td>\n",
       "      <td>We sought to identify fibroblast growth factor...</td>\n",
       "      <td>7</td>\n",
       "      <td>8619</td>\n",
       "      <td>we sought to identify fibroblast growth factor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>5199</td>\n",
       "      <td>ALK</td>\n",
       "      <td>F1174C</td>\n",
       "      <td>In the era of personalized medicine, understan...</td>\n",
       "      <td>7</td>\n",
       "      <td>11175</td>\n",
       "      <td>ization of potent and selective second generat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>5213</td>\n",
       "      <td>MET</td>\n",
       "      <td>X1006_splice</td>\n",
       "      <td>Non–small cell lung cancer (NSCLC) is a diffic...</td>\n",
       "      <td>7</td>\n",
       "      <td>15035</td>\n",
       "      <td>ed death rate of 160,440 cases in the united s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>5219</td>\n",
       "      <td>KEAP1</td>\n",
       "      <td>G333C</td>\n",
       "      <td>IκB kinase β (IKKβ) is involved in tumor devel...</td>\n",
       "      <td>4</td>\n",
       "      <td>3643</td>\n",
       "      <td>tribute to cancer development via the up-regul...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>5241</td>\n",
       "      <td>TP53</td>\n",
       "      <td>V272M</td>\n",
       "      <td>The transcription factor and tumor suppressor ...</td>\n",
       "      <td>1</td>\n",
       "      <td>8794</td>\n",
       "      <td>of the suppressor mechanism.  understanding th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>5266</td>\n",
       "      <td>PDGFRA</td>\n",
       "      <td>CDK5RAP2-PDGFRA Fusion</td>\n",
       "      <td>Chronic myeloproliferative disorders with rear...</td>\n",
       "      <td>7</td>\n",
       "      <td>2640</td>\n",
       "      <td>osine kinase-encoding genes at translocation b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346</th>\n",
       "      <td>5316</td>\n",
       "      <td>GATA3</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Terminal deletions of chromosome 10p result in...</td>\n",
       "      <td>1</td>\n",
       "      <td>21901</td>\n",
       "      <td>terminal deletions of chromosome 10p result in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347</th>\n",
       "      <td>5341</td>\n",
       "      <td>HNF1A</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Mutations in the hepatocyte nuclear factor (HN...</td>\n",
       "      <td>1</td>\n",
       "      <td>10283</td>\n",
       "      <td>dna binding because both mutations are positi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>5351</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>Q257R</td>\n",
       "      <td>The Ras/MAPK pathway is critical for human dev...</td>\n",
       "      <td>7</td>\n",
       "      <td>8220</td>\n",
       "      <td>on about the treatment strategy for individual...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>5405</td>\n",
       "      <td>PTPN11</td>\n",
       "      <td>R498W</td>\n",
       "      <td>Elucidation of the molecular mechanisms underl...</td>\n",
       "      <td>1</td>\n",
       "      <td>5367</td>\n",
       "      <td>l table link at the top of the online article)...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>5409</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>R130Q</td>\n",
       "      <td>The tumor suppressor gene PTEN is frequently m...</td>\n",
       "      <td>4</td>\n",
       "      <td>3971</td>\n",
       "      <td>r was constructed as follows. the full-length ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>5426</td>\n",
       "      <td>FLT3</td>\n",
       "      <td>K602_W603insYEYDLK</td>\n",
       "      <td>An internal tandem duplication (ITD) of the FL...</td>\n",
       "      <td>7</td>\n",
       "      <td>14829</td>\n",
       "      <td>an internal tandem duplication (itd) of the fl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>5436</td>\n",
       "      <td>FGFR3</td>\n",
       "      <td>I538F</td>\n",
       "      <td>Frequent genetic alterations discovered in FGF...</td>\n",
       "      <td>5</td>\n",
       "      <td>5499</td>\n",
       "      <td>frequent genetic alterations discovered in fgf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>5437</td>\n",
       "      <td>CCND3</td>\n",
       "      <td>D286_L292trunc</td>\n",
       "      <td>The activities of cyclin D-dependent kinases s...</td>\n",
       "      <td>2</td>\n",
       "      <td>6376</td>\n",
       "      <td>d recovered immune complexes were incubated in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>5451</td>\n",
       "      <td>AKT2</td>\n",
       "      <td>BCAM-AKT2 Fusion</td>\n",
       "      <td>High-grade serous ovarian cancer (HGSC) is the...</td>\n",
       "      <td>7</td>\n",
       "      <td>3651</td>\n",
       "      <td>high-grade serous ovarian cancer (hgsc) is the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>5464</td>\n",
       "      <td>KIT</td>\n",
       "      <td>V560del</td>\n",
       "      <td>To analyze a multi-institutional series of typ...</td>\n",
       "      <td>7</td>\n",
       "      <td>17259</td>\n",
       "      <td>of 17 tcs, but no mutations were detected in ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>5487</td>\n",
       "      <td>RET</td>\n",
       "      <td>E511K</td>\n",
       "      <td>Context Most germline-activating mutations of ...</td>\n",
       "      <td>2</td>\n",
       "      <td>8905</td>\n",
       "      <td>context most germline-activating mutations of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>5500</td>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>G34A</td>\n",
       "      <td>We screened 75 primary hepatocellular carcinom...</td>\n",
       "      <td>7</td>\n",
       "      <td>9109</td>\n",
       "      <td>s in the same gene may reflect different etiol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>5509</td>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>D32G</td>\n",
       "      <td>We screened 75 primary hepatocellular carcinom...</td>\n",
       "      <td>7</td>\n",
       "      <td>8858</td>\n",
       "      <td>gene may reflect different etiologies of carci...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>5515</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>P1812A</td>\n",
       "      <td>Abstract  The BRCA1 gene from individuals at ...</td>\n",
       "      <td>1</td>\n",
       "      <td>3731</td>\n",
       "      <td>19, 20, 21) and supported by the mutation set ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>5531</td>\n",
       "      <td>NFE2L2</td>\n",
       "      <td>D29N</td>\n",
       "      <td>The Nrf2 (nuclear factor erythroid 2 [NF-E2]-r...</td>\n",
       "      <td>7</td>\n",
       "      <td>9876</td>\n",
       "      <td>oresistance but also acquired resistance in ca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>5564</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>G129D</td>\n",
       "      <td>The PTEN (phosphatase and tensin homolog) phos...</td>\n",
       "      <td>4</td>\n",
       "      <td>6351</td>\n",
       "      <td>ion to rescue growth arrest induced by p110α-c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>5581</td>\n",
       "      <td>MAP2K1</td>\n",
       "      <td>L115P</td>\n",
       "      <td>Genetic alterations that activate the mitogen-...</td>\n",
       "      <td>2</td>\n",
       "      <td>3541</td>\n",
       "      <td>from mutatgenesis experiments using ci-1040 (f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>5600</td>\n",
       "      <td>MAP2K4</td>\n",
       "      <td>S251N</td>\n",
       "      <td>The systematic characterization of somatic mut...</td>\n",
       "      <td>4</td>\n",
       "      <td>8073</td>\n",
       "      <td>the systematic characterization of somatic mut...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>5609</td>\n",
       "      <td>VHL</td>\n",
       "      <td>N90I</td>\n",
       "      <td>The von Hippel-Lindau tumor suppressor protein...</td>\n",
       "      <td>4</td>\n",
       "      <td>7082</td>\n",
       "      <td>howed much reduced (y112h) or absent co-precip...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>5626</td>\n",
       "      <td>MAP2K1</td>\n",
       "      <td>H119P</td>\n",
       "      <td>Genetic alterations that activate the mitogen-...</td>\n",
       "      <td>6</td>\n",
       "      <td>3541</td>\n",
       "      <td>mutation and are highly sensitive to mek inhib...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>5628</td>\n",
       "      <td>ATM</td>\n",
       "      <td>V2424G</td>\n",
       "      <td>Abstract The functional consequences of misse...</td>\n",
       "      <td>4</td>\n",
       "      <td>11621</td>\n",
       "      <td>ll other constructs with missense variants. un...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>5650</td>\n",
       "      <td>TP53</td>\n",
       "      <td>G245C</td>\n",
       "      <td>The transcription factor and tumor suppressor ...</td>\n",
       "      <td>1</td>\n",
       "      <td>4351</td>\n",
       "      <td>40. this finding could theoretically reflect v...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>367 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID     Gene               Variation  \\\n",
       "0      12     TET2                  Y1902A   \n",
       "1      19     MTOR                  D2512H   \n",
       "2      21      KIT                    D52N   \n",
       "3      55     SPOP                   F125V   \n",
       "4      64    KEAP1                    C23Y   \n",
       "5      70   CDKN2A                    G35V   \n",
       "6      89     IDH1                   R132S   \n",
       "7      93      KIT                   N822K   \n",
       "8     111     ETV1           Amplification   \n",
       "9     113     JAK1                   F958C   \n",
       "10    119      KIT            P551_M552del   \n",
       "11    121    BRCA1                  T1773I   \n",
       "12    126   TGFBR2                   E526Q   \n",
       "13    137     TP53                   Q144P   \n",
       "14    159    BRCA2    Truncating Mutations   \n",
       "15    215    BRCA2                   G173C   \n",
       "16    226     GNAQ                   R183Q   \n",
       "17    234     MTOR                  R2152C   \n",
       "18    286   PIK3CA                   E453Q   \n",
       "19    344   PDGFRA            W559_R560del   \n",
       "20    353     EGFR                   S720P   \n",
       "21    381  PPP2R1A                   R183W   \n",
       "22    382    EWSR1        EWSR1-ERG Fusion   \n",
       "23    426    ERBB4                   D931Y   \n",
       "24    469     TSC1                   R509Q   \n",
       "25    535     FLT3                   D835V   \n",
       "26    542    SF3B1                   H662D   \n",
       "27    545     BRAF                   F595L   \n",
       "28    552   PIK3R1    Truncating Mutations   \n",
       "29    562     TP53                   E180K   \n",
       "..    ...      ...                     ...   \n",
       "338  5153      SMO                   W281C   \n",
       "339  5178     TP53                   C176R   \n",
       "340  5192    FGFR2                   N549H   \n",
       "341  5199      ALK                  F1174C   \n",
       "342  5213      MET            X1006_splice   \n",
       "343  5219    KEAP1                   G333C   \n",
       "344  5241     TP53                   V272M   \n",
       "345  5266   PDGFRA  CDK5RAP2-PDGFRA Fusion   \n",
       "346  5316    GATA3                Deletion   \n",
       "347  5341    HNF1A                Deletion   \n",
       "348  5351     BRAF                   Q257R   \n",
       "349  5405   PTPN11                   R498W   \n",
       "350  5409     PTEN                   R130Q   \n",
       "351  5426     FLT3      K602_W603insYEYDLK   \n",
       "352  5436    FGFR3                   I538F   \n",
       "353  5437    CCND3          D286_L292trunc   \n",
       "354  5451     AKT2        BCAM-AKT2 Fusion   \n",
       "355  5464      KIT                 V560del   \n",
       "356  5487      RET                   E511K   \n",
       "357  5500   CTNNB1                    G34A   \n",
       "358  5509   CTNNB1                    D32G   \n",
       "359  5515    BRCA1                  P1812A   \n",
       "360  5531   NFE2L2                    D29N   \n",
       "361  5564     PTEN                   G129D   \n",
       "362  5581   MAP2K1                   L115P   \n",
       "363  5600   MAP2K4                   S251N   \n",
       "364  5609      VHL                    N90I   \n",
       "365  5626   MAP2K1                   H119P   \n",
       "366  5628      ATM                  V2424G   \n",
       "367  5650     TP53                   G245C   \n",
       "\n",
       "                                                  Text  Class  Text_count  \\\n",
       "0    TET proteins oxidize 5-methylcytosine (5mC) on...      1        7123   \n",
       "1    Genes encoding components of the PI3K-Akt-mTOR...      2        4283   \n",
       "2    Myeloproliferative disorders (MPD) constitute ...      2        2477   \n",
       "3    In the largest E3 ligase subfamily, Cul3 binds...      4       22727   \n",
       "4    Keap1 is the substrate recognition module of a...      4        2854   \n",
       "5    Inherited mutations affecting the INK4a/ARF lo...      4       12660   \n",
       "6    To assess the value of anti-isocitrate dehydro...      9       27549   \n",
       "7    c-KIT is a member of the type 3 subclass of tr...      7       34300   \n",
       "8    E26 transformation-specific (ETS) transcriptio...      7        7082   \n",
       "9    Activating mutations in JAK1 and JAK2 have bee...      7        4878   \n",
       "10   The pathogenesis of most gastrointestinal stro...      2        3997   \n",
       "11   Genetic screening of the breast and ovarian ca...      1        5245   \n",
       "12   Transforming growth factor β (TGFβ) plays a ke...      4        9016   \n",
       "13   Inheritance of germ-line mutant alleles of BRC...      1        4635   \n",
       "14    Abstract Heterozygous germ line mutations in ...      1       25258   \n",
       "15   Exonic variants of unknown biological signific...      1        5883   \n",
       "16   This study represents the most comprehensive m...      2        3439   \n",
       "17   Genes encoding components of the PI3K-Akt-mTOR...      3        4283   \n",
       "18   Mutations in the PIK3CA gene, which encodes th...      2        6877   \n",
       "19   Gastrointestinal stromal tumors (GISTs) are co...      7        1836   \n",
       "20   The epidermal growth factor receptor tyrosine ...      2        6880   \n",
       "21   Protein Phosphatase 2A (PP2A) plays an essenti...      1       12291   \n",
       "22   Ets-related gene (ERG), which encodes a member...      7        9050   \n",
       "23   Recent efforts to comprehensively characterize...      7        5337   \n",
       "24   Tuberous sclerosis complex (TSC) is an autosom...      3        3820   \n",
       "25   Mutations of receptor tyrosine kinases are imp...      7       10694   \n",
       "26   The splicing factor SF3B1 is the most commonly...      9        5517   \n",
       "27   Recently, mutations in the B-Raf gene have bee...      7       11983   \n",
       "28   Class IA phosphatidylinositol 3-kinases (PI3K)...      1       15901   \n",
       "29   Under conditions of genotoxic stress, human p5...      4       12107   \n",
       "..                                                 ...    ...         ...   \n",
       "338  Introduction Uncontrolled activation of the He...      7       13821   \n",
       "339  More than half of all human cancers are associ...      4        3069   \n",
       "340  We sought to identify fibroblast growth factor...      7        8619   \n",
       "341  In the era of personalized medicine, understan...      7       11175   \n",
       "342  Non–small cell lung cancer (NSCLC) is a diffic...      7       15035   \n",
       "343  IκB kinase β (IKKβ) is involved in tumor devel...      4        3643   \n",
       "344  The transcription factor and tumor suppressor ...      1        8794   \n",
       "345  Chronic myeloproliferative disorders with rear...      7        2640   \n",
       "346  Terminal deletions of chromosome 10p result in...      1       21901   \n",
       "347  Mutations in the hepatocyte nuclear factor (HN...      1       10283   \n",
       "348  The Ras/MAPK pathway is critical for human dev...      7        8220   \n",
       "349  Elucidation of the molecular mechanisms underl...      1        5367   \n",
       "350  The tumor suppressor gene PTEN is frequently m...      4        3971   \n",
       "351  An internal tandem duplication (ITD) of the FL...      7       14829   \n",
       "352  Frequent genetic alterations discovered in FGF...      5        5499   \n",
       "353  The activities of cyclin D-dependent kinases s...      2        6376   \n",
       "354  High-grade serous ovarian cancer (HGSC) is the...      7        3651   \n",
       "355  To analyze a multi-institutional series of typ...      7       17259   \n",
       "356  Context Most germline-activating mutations of ...      2        8905   \n",
       "357  We screened 75 primary hepatocellular carcinom...      7        9109   \n",
       "358  We screened 75 primary hepatocellular carcinom...      7        8858   \n",
       "359   Abstract  The BRCA1 gene from individuals at ...      1        3731   \n",
       "360  The Nrf2 (nuclear factor erythroid 2 [NF-E2]-r...      7        9876   \n",
       "361  The PTEN (phosphatase and tensin homolog) phos...      4        6351   \n",
       "362  Genetic alterations that activate the mitogen-...      2        3541   \n",
       "363  The systematic characterization of somatic mut...      4        8073   \n",
       "364  The von Hippel-Lindau tumor suppressor protein...      4        7082   \n",
       "365  Genetic alterations that activate the mitogen-...      6        3541   \n",
       "366   Abstract The functional consequences of misse...      4       11621   \n",
       "367  The transcription factor and tumor suppressor ...      1        4351   \n",
       "\n",
       "                                         modified_text  \n",
       "0     the binding of purified tet2 proteins to a fa...  \n",
       "1    red somatic point mutations (supplementary tab...  \n",
       "2    myeloproliferative disorders (mpd) constitute ...  \n",
       "3    re molecular signatures database (msigdb, cont...  \n",
       "4    keap1 is the substrate recognition module of a...  \n",
       "5    inherited mutations affecting the ink4a/arf lo...  \n",
       "6    09 was generally crisper with a better signal-...  \n",
       "7    us 10% fetal calf serum (fcs), and 4 mg cell l...  \n",
       "8     to be fully explored. we have completed a com...  \n",
       "9    pocket formed between the jh2 αc helix and an ...  \n",
       "10   the pathogenesis of most gastrointestinal stro...  \n",
       "11   lysis to assess the protein folding stability ...  \n",
       "12   β) plays a key role in maintaining tissue home...  \n",
       "13   inheritance of germ-line mutant alleles of brc...  \n",
       "14   ers of the family (two sisters) who were alrea...  \n",
       "15   exonic variants of unknown biological signific...  \n",
       "16   ell et al., 1989; luyten et al., 1996; manioti...  \n",
       "17    parental or wt-expressing cells (figure 3b an...  \n",
       "18   0a expression vector pcmv6-ish2p110a were kind...  \n",
       "19   gastrointestinal stromal tumors (gists) are co...  \n",
       "20   the epidermal growth factor receptor tyrosine ...  \n",
       "21   s, the pp2a core enzyme associates with a vari...  \n",
       "22   ets-related gene (erg), which encodes a member...  \n",
       "23   recent efforts to comprehensively characterize...  \n",
       "24   y the complex and on the phosphorylation statu...  \n",
       "25   mutations of receptor tyrosine kinases are imp...  \n",
       "26   the splicing factor sf3b1 is the most commonly...  \n",
       "27   recently, mutations in the b-raf gene have bee...  \n",
       "28   class ia phosphatidylinositol 3-kinases (pi3k)...  \n",
       "29   rs35,36. we also mutated corresponding interac...  \n",
       "..                                                 ...  \n",
       "338   the pocket without directly contacting the in...  \n",
       "339  mi) for the macintosh. the core domains ofadh-...  \n",
       "340  we sought to identify fibroblast growth factor...  \n",
       "341  ization of potent and selective second generat...  \n",
       "342  ed death rate of 160,440 cases in the united s...  \n",
       "343  tribute to cancer development via the up-regul...  \n",
       "344  of the suppressor mechanism.  understanding th...  \n",
       "345  osine kinase-encoding genes at translocation b...  \n",
       "346  terminal deletions of chromosome 10p result in...  \n",
       "347   dna binding because both mutations are positi...  \n",
       "348  on about the treatment strategy for individual...  \n",
       "349  l table link at the top of the online article)...  \n",
       "350  r was constructed as follows. the full-length ...  \n",
       "351  an internal tandem duplication (itd) of the fl...  \n",
       "352  frequent genetic alterations discovered in fgf...  \n",
       "353  d recovered immune complexes were incubated in...  \n",
       "354  high-grade serous ovarian cancer (hgsc) is the...  \n",
       "355   of 17 tcs, but no mutations were detected in ...  \n",
       "356  context most germline-activating mutations of ...  \n",
       "357  s in the same gene may reflect different etiol...  \n",
       "358  gene may reflect different etiologies of carci...  \n",
       "359  19, 20, 21) and supported by the mutation set ...  \n",
       "360  oresistance but also acquired resistance in ca...  \n",
       "361  ion to rescue growth arrest induced by p110α-c...  \n",
       "362  from mutatgenesis experiments using ci-1040 (f...  \n",
       "363  the systematic characterization of somatic mut...  \n",
       "364  howed much reduced (y112h) or absent co-precip...  \n",
       "365  mutation and are highly sensitive to mek inhib...  \n",
       "366  ll other constructs with missense variants. un...  \n",
       "367  40. this finding could theoretically reflect v...  \n",
       "\n",
       "[367 rows x 7 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_classified_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_classified_df[['Gene', 'Variation']].to_pickle('test_variants_filtered')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "colab": {
     "autoexec": {
      "startup": false,
      "wait_interval": 0
     }
    },
    "colab_type": "code",
    "id": "qaEgQ67KVsPW"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(367,)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_test=test_classified_df[['Text', 'Gene','Variation', 'ID']].apply(lambda x: extract_text_sections(x[0].lower(),x[1].lower(),x[2].lower(), x[3]), axis=1 )\n",
    "X_test.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "colab": {
     "autoexec": {
      "startup": false,
      "wait_interval": 0
     }
    },
    "colab_type": "code",
    "id": "S9spEEVSYXOn"
   },
   "outputs": [],
   "source": [
    "#X_test=test_classified_df['Text']\n",
    "Y_test = test_classified_df['Class']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_classified_df.to_pickle('test_1000.sav')\n",
    "#Y_test.to_pickle('Y_test_1000.sav')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "8WyDHKjybExg"
   },
   "source": [
    "# Saving the training and test set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "autoexec": {
      "startup": false,
      "wait_interval": 0
     }
    },
    "colab_type": "code",
    "id": "S5B9Y_n5WNIz"
   },
   "outputs": [],
   "source": [
    "# Saving it temporary on the computer\n",
    "X_train.to_pickle('/tmp/X_train_1000.sav')\n",
    "Y_train.to_pickle('/tmp/Y_train_1000.sav')\n",
    "X_test.to_pickle('/tmp/X_test_1000.sav')\n",
    "Y_test.to_pickle('/tmp/Y_test_1000.sav')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "autoexec": {
      "startup": false,
      "wait_interval": 0
     }
    },
    "colab_type": "code",
    "id": "fiTmONCBW5J_"
   },
   "outputs": [],
   "source": [
    "pickled_df = {'/tmp/X_train_1000.sav':'X_train_1000', '/tmp/Y_train_1000.sav':'Y_train_1000', '/tmp/X_test_1000.sav': 'X_test_1000', '/tmp/Y_test_1000.sav':'Y_test_1000' }"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "default_view": {},
   "name": "Training_and_Test_files.ipynb",
   "private_outputs": true,
   "provenance": [],
   "version": "0.3.2",
   "views": {}
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
